File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
import heapq
# 최소비용정리,

def solution(n, costs):
    # n : 섬의 개수 
    graph = [[] for _ in range(n) ]
    for cost in costs :
        island1, island2, c = cost 
        graph [island1].append ((c,island2))
        graph [island2].append ((c,island1))
    
    
    # 다익스트라
    # 모든섬 통행 -> 최소비용
    def prim (start) :
        visited = [False] *(n)
        cnt = 0
        total_cost = 0
        q = [(0,start)]
        
        while q :
            cur_cost, cur_node = heapq.heappop(q) 
            if visited[cur_node]  : #방문한 곳이라면? 스킵
                continue
            visited[cur_node] = True
            total_cost = total_cost + cur_cost
            cnt +=1
            if cnt == n :
                return total_cost
            for nxt_cost, nxt_node in graph [cur_node] :
                if not visited[nxt_node] :
                    heapq.heappush (q,(nxt_cost,nxt_node))
            
            
    
        
    return prim(0)

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] costs) {
        
        ArrayList<int[]>[] graph = new ArrayList [n];
        for (int i =0; i<n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for(int [] item: costs) {
            int start = item[0];
            int end = item[1];
            int cost = item[2];
            
            graph[start].add(new int [] {cost,end});
            graph[end].add(new int [] {cost,start});
            
        }
        
        
        PriorityQueue <int []> q = new PriorityQueue <>((a,b) -> a[0] -b[0]);
        
        q.offer(new int[] {0,0});
        int cnt = 0;
        int answer =0;
        
        boolean[] visited = new boolean[n];
        int total_cost = 0;
        Arrays.fill(visited, false);
        while (!q.isEmpty()) {
            int []item = q.poll();            
            int cur_cost =  item[0];
            int cur_node =  item[1];
            
            if(visited[cur_node]) {
               continue; 
            }
            
         
            
            visited[cur_node] = true;
            total_cost = total_cost + cur_cost;
            cnt+=1;
            
            if (cnt == n) {
                answer = total_cost;
                break;
            }
            for (int [] nxt : graph[cur_node]) {
                int nxt_cost = nxt[0];
                int nxt_node = nxt[1];
                
                if (!visited[nxt_node]) {
                    q.offer (new int [] {nxt_cost, nxt_node});
                }
            }
            
        }
        
        return answer;
        
    }
}